### PR TITLE
Generalize GenerateMissingRecordFieldsFix to work on functions too.

### PR DESCRIPTION
**Pull request by mcon**
# Current behaviour
```
type R = { A: int; B: int }
let myVariable : R = {}
```
to
```
type R = { A: int; B: int }
let myVariable : R = { A = failwith "todo"
                       B = failwith "todo" }
```
 - Only works with variable bindings
# Enhancement: work with function bindings too
e.g.
```
type R = { A: int; B: int }
let myFunction a b : R = {}
```
to
```
type R = { A: int; B: int }
let myFunction a b : R = { A = failwith "todo"
                           B = failwith "todo" }
```

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs [hunk 1]
```diff
@@ -31,6 +32,15 @@ type GenerateMissingRecordFieldsFix(recordExpr: IRecordExpr) =
             | null -> failwith "Could not get expr"
             | expr -> ModificationUtil.AddChildAfter(expr, FSharpTokenType.SEMICOLON.CreateLeafElement()) |> ignore
 
+    let getReturnType (recordExpr : IRecordExpr) =
+        match recordExpr.SelfAndPathToRoot() |> Seq.tryPick(fun x -> x.As<ITopBinding>() |> Option.ofObj) with
```
[the remark this answers is not in the document]
Comment (mcon) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs:36]: I should have known it wouldn't be this easy! Thanks for the quick feedback @auduchinok

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs [hunk 2]
```diff
@@ -31,6 +32,15 @@ type GenerateMissingRecordFieldsFix(recordExpr: IRecordExpr) =
...
+        | None -> failwith "Could not get top binding reference"
+        | Some topBinding ->
+        match topBinding.DeclaredElement.As<IFSharpMember>() with
```
[the remark this answers is not in the document]
Comment (mcon) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs]: I couldn't quite work this one out I'm afraid.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcon) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs]: As suggested I've created a `getReturnExpressionOwner` and also added a new testcase to cover the case you've pointed out.

In order to get things working, I needed to make `IBinding` implement `IFunctionDeclaration`, which seems quite reasonable, though the implementation is now duplicated between `TopBinding` and `LocalBinding`. I haven't looked into how much effort unifying the two of these would be.

Let me know if you'd like me to move the `getReturnExpressionOwner` logic into `RecordExpr` as part of this PR.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Util/FSharpExpressionUtil.fs [hunk 3]
```diff
@@ -12,6 +12,26 @@ open JetBrains.ReSharper.Psi
 open JetBrains.ReSharper.Psi.ExtensionsAPI.Tree
 open JetBrains.ReSharper.Psi.Tree
 
+/// If given an expression is a return expression, the owning IBinding is returned, otherwise null.
+let getReturnExpressionOwner (expr: ISynExpr) : IBinding =
+    let rec getParentFunction (treeNode : ITreeNode) =
+        let parent = treeNode.Parent
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/Util/FSharpExpressionUtil.fs:18]: I'd probably expect it to be `ISynExpr -> ISynExpr` (though, `ISynExpr` was recently renamed to `IFSharpExpression` after the FCS-inspired name produced many questions).

Getting an outer expression that is a parent to some return expression is needed in various places and is not limited to `let`-bindings. Speaking of top-level declarations, it can also be used for auto properties or other members in types. Other cases, for instance, include case where it'd help to reason about a type of an `if` expression by looking at its return expression in `if` branch.

This code should really cover common cases via generated navigators like:
`IfThenElseExprNavigator.GetByBranchExpression`
`MatchClauseNavigator.GetByExpression`
`MatchLikeExprNavigator.GetByClause`
and so on.

In most cases simply taking a parent is wrong (and this is why it wasn't written this way with a `todo` to cover it properly):
```fsharp
match 123 with
| x when { } -> { }
```
In the code above there are multiple expressions including two unfinished record ones. Only the second of this two is considered a "return" expression for the match expression.

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/LocalBinding.cs [hunk 4]
```diff
@@ -18,5 +24,33 @@ public void SetIsMutable(bool value)
       if (headPat != null)
         FSharpImplUtil.AddTokenBefore(headPat, FSharpTokenType.MUTABLE);
     }
+
+    /// A workaround for getting a declared element for binding in features like Find Usages results and
+    /// file member navigation where we're looking for containing type member.
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/LocalBinding.cs:29]: The binding types currently don't implement it due to it'd be more difficult to work with in cases with multile declared elements like:
```fsharp
let x, y = 1, 2
let (UnionCase(fieldA, fieldB) = Foo
```
Please don't add these members to binding types, at least of a part of this PR where it seems unrelated.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs [hunk 5]
```diff
@@ -31,6 +33,20 @@ type GenerateMissingRecordFieldsFix(recordExpr: IRecordExpr) =
             | null -> failwith "Could not get expr"
             | expr -> ModificationUtil.AddChildAfter(expr, FSharpTokenType.SEMICOLON.CreateLeafElement()) |> ignore
 
+    let getReturnRecordFieldNames (recordExpr : IRecordExpr) : Option<seq<string>> =
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs:36]: Please don't duplicate `GetRecordFieldNames`.

**Comment by mcon** [comment]
Sorry for misunderstanding your comments first time round - I didn't appreciate how `RecordExpr.GetFSharpSymbol()` was related to everything before.

I'm going to do `match`, `function` and `if` statements now - up to you whether that's part of this PR or a new one.

**Comment by auduchinok** [comment]
@mcon Thank you, it's indeed much cleaner now!
I think it'd be good to get it merged and to work on the other construct in separate PRs. Should I merge it if you remove the draft mark and everything rest seems fine?

**Comment by auduchinok** [comment]
@mcon Thanks!
